### ISSUE TITLE
Insert a Date in to mssql

### ISSUE THREAD

**Issue by taoqf** [knex/knex#6108]
# Environment

Knex version:2.5.1
Database + version:mssql
OS:Debian

Select applicable template from below.
If issue is about oracledb support, tag @ atiertant. For MSSql tag @ smorey2.
Rest of dialects don't need tags.
If issue is about TypeScript definitions, tag @ lorefnon.
@smorey2
# Bug

1. Inser an object value
{
dtfield: new Date()
}
2. value inserted but without timezone But if I do a query with Date, it will has timezone dealed. I checked the source code, could not find a clue.

**Comment by DanielVenable** [comment]
There are several different date fields in mssql, but it looks like datetimeoffset is the only field that contains time zone information. (See [this article](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql?view=sql-server-ver16)). Is that what your database is using?

**Comment by taoqf** [comment]
I'm using `datetime` .

**Comment by taoqf** [comment]
@DanielVenable

**Comment by mercmobily** [comment]
This is actually expected behaviour based on how SQL Server stores dates and how the underlying Tedious driver handles them.

The behaviour you're describing is a fundamental characteristic of SQL Server's datetime types and how JavaScript Date objects interact with them. When you insert a JavaScript Date object into an MSSQL datetime column, Knex uses the Tedious driver which maps JavaScript Date values to the `TYPES.DateTime` type (you can see this in lib/dialects/mssql/index.js:375). The issue is that SQL Server's `datetime` and `datetime2` types do not store timezone information—they only store the raw date and time values.

However, when the Tedious driver sends a JavaScript Date to SQL Server, it applies timezone conversions inconsistently. As documented in several Tedious issues (particularly tediousjs/tedious#603 and tediousjs/tedious#675), the driver applies timezone offset adjustments when writing DateTime values but not when reading them back, creating exactly the discrepancy you've observed. This is compounded by the fact that the `useUTC` configuration option in Tedious only affects `datetime` and `smalldatetime` types, not `datetime2` types, though this has been improved in later versions of Tedious.

If you genuinely need to preserve timezone information (not just UTC timestamps), you should use SQL Server's `datetimeoffset` type, which is the only MSSQL type that actually stores timezone offset data alongside the datetime value. You can create this column type in Knex using:

```javascript
table.specificType('dtfield', 'datetimeoffset')
```

As elhigu explained in issue #2501, even PostgreSQL's `timestamptz` doesn't actually store timezone information—it just converts values to UTC during storage. SQL Server's `datetimeoffset` is one of the few database types that genuinely preserves timezone offsets.

Alternatively, if you're storing UTC timestamps (which is the recommended approach for most applications), you can configure Tedious to use UTC consistently by adding `useUTC: true` to your connection options:

```javascript
{
  client: 'mssql',
  connection: {
    // ... other config
    options: {
      useUTC: true
    }
  }
}
```

This should help ensure more consistent timezone handling, though be aware that the `useUTC` option has had various bugs and behavioural changes across different Tedious versions.

The root issue here is that Knex relies on the Tedious driver's type mapping, and Tedious automatically uses `TYPES.DateTime` for JavaScript Date objects. Whilst Knex now uses `datetime2` for schema migrations (since PR #2757 based on Microsoft's recommendations in issue #2304), this only affects column definitions—it doesn't change how bound Date parameters are sent to the database during inserts and queries.

In summary, this is not a bug in Knex but rather a consequence of how SQL Server's datetime types work and how Tedious handles timezone conversions. For applications that need timezone information, use `datetimeoffset`. For applications storing UTC timestamps, ensure your application always works in UTC and consider setting `useUTC: true` in your Tedious options.


Related issues: #2501, #2304, #5653, #6017
